feat(addition-of-a-new-domain-and-guide-tracking-labwork-and-marking)… (#64)
btpworks/chronicle-examples: domains/scienceproject/guide.md

# Science Project Guide

## Modelling a University Science Project

> Academic dishonesty is unfortunately rife in Universities. Here we show a

Reviewer: ```suggestion
Academic dishonesty is unfortunately rife in universities. Here we show a
```

> chronicle domain that can be used to track the activities of a student and the

Reviewer: ```suggestion
Chronicle domain that can be used to track the activities of a student and the
```

subsequent marking by the university professor

> ### Modellng Labwork

Reviewer: ```suggestion
### Modelling Labwork
```

#### Modelling the Student Agent

The `Student` agent has two attributes, its `StudentID` and its `Grade`.
In our Chronicle domain specification this is captured as follows -

```yaml
agents:
  Student:
    attributes:
      - StudentID
      - Grade
attributes:
  StudentID:
    type: "String"
  Grade:
    type: "String"
```

#### Modelling the Experiment Entity

The `Experiment` entity has three attributes, its `Reaction`, the `Reagents`
used in the reaction, and the `Approval` status of the reaction.

```yaml
  Experiment:
    attributes:
      - Reaction
      - Reagents
      - Approval
attributes:
  Reaction:

>     type: "String"

Reviewer: ```suggestion
    type: String
```

  Reagents:
    type: "String"
  Approval:
    type: "String"
```

### Modelling the Labwork Activity

The `Labwork` activity has three attributes, the `StudentID` doing the labwork,
the `Workplan` of the labwork, and the `Approval` status of the Labwork.

```yaml
  Labwork:
    attributes:
      - StudentID
      - Workplan
      - Approval
attributes:
    StudentID:

>         type: "String"

Reviewer: ```suggestion
        type: String
```

    Workplan:
        type: "String"
    Approval:
        type: "String"

```

> ### Modeling the results of Labwork

Reviewer: ```suggestion
### Modeling the results of Labwork
```
I would actually go with US spellings (modeling rather than modelling) in this context but consistency more so.

> the students lab-work generates the `Results` entity that has 4 attributes. The

Reviewer: ```suggestion
the student's lab-work generates the `Results` entity that has 4 attributes. The
```

`RawData`, the `ReportURL`, the `Purity` of the chemistry produced, and the
`Yield`.

```yaml
Results:
  attributes:
    - RawData
    - ReportURL
    - Purity
    - Yield
```

> ### Assesment of Labwork

Reviewer: ```suggestion
### Assessment of Labwork
```

In this Model the Labwork is being assessed by a Professor

> #### Modelling the Marks Entity

Reviewer: ```suggestion
#### Modeling the Marks Entity
```

The `Marks` entity has three attributes, its `Grade`, the total `Marks` awarded,
and a `FeedbackURL`.

```yaml
entities:
  Marks:
    attributes:
      - Grade
      - Marks
      - FeedbackURL
```

> #### Modelling the Assesment Activity

Reviewer: ```suggestion
#### Modelling the Assessment Activity
```

The `Assessment` activity has two attributes the `StaffID` of the assessor and
the `ReportURL` being assessed.

```yaml
activities:
  Assessment:
    attributes:
      - StaffID
      - ReportURL
```

### Chronicle Domain

Combining these fragments gives us our Chronicle `scienceproject` domain.

```yaml
name: "scienceproject"
agents:
  Student:
    attributes:
      - StudentID
      - Grade
  Professor:
    attributes:
      - StaffID
      - Department
entities:
  Experiment:
    attributes:
      - Reaction
      - Reagents
      - Approval
  Results:
    attributes:
      - RawData
      - ReportURL
      - Purity
      - Yield
  Marks:
    attributes:
      - Grade
      - Marks
      - FeedbackURL
activities:
  Labwork:
    attributes:
      - StudentID
      - Workplan
      - Approval
  Assessment:
    attributes:
      - StaffID
      - ReportURL
roles:
  - student
  - professor
attributes:
  StudentID:

>     type: "String"

Reviewer: ```suggestion
    type: String
```

  Grade:
    type: "String"
  StaffID:
    type: "String"
  Department:
    type: "String"
  Reaction:
    type: "String"
  Reagents:
    type: "String"
  Approval:
    type: "String"
  RawData:
    type: "String"
  ReportURL:
    type: "String"
  Purity:
    type: "String"
  Yield:
    type: "String"
  Marks:
    type: "String"
  FeedbackURL:
    type: "String"
  Workplan:
    type: "String"
```

> ## Recording the Science project

Reviewer: ```suggestion
## Recording the Science Project
```

> In this example we will create a 'Student' and 'Professor' agent. These agents

Reviewer: ```suggestion
In this example we will create a `Student` and `Professor` agent. These agents
```

will complete the following activities:

> 1. "Labwork" activity to record when a student has completed some labwork to be

Reviewer: ```suggestion
1. `Labwork` activity to record when a student has completed some labwork to be
```

   graded.

> 1. "Assesment" an activity to record when a professor has assesed a students

Reviewer: ```suggestion
1. "Assessment" an activity to record when a professor has assessed a student's
```

   labwork.

> ### Defining agents

Reviewer: ```suggestion
### Defining Agents
```

Here is where we define who is involved in this example. In a real world example

> there would likley be many more students and a few more professor agents.

Reviewer: ```suggestion
there would likely be many more students and a few more professor agents.
```

```graphql
mutation defineAgents {
  defineStudentAgent(
    externalId: "Joe BLoggs"
    attributes: { studentIDAttribute: "Joe Blogs", gradeAttribute: "A" }
  ) {
    context
    txId
  }
  defineProfessorAgent(
    externalId: "Prof Plum"
    attributes: {
      staffIDAttribute: "Prof Plum"
      departmentAttribute: "Chemistry"
    }
  ) {
    context
    txId
  }
}
```

Which will output something similar to

```json
{
  "data": {
    "defineStudentAgent": {
      "context": "chronicle:agent:Joe%20BLoggs",
      "txId": "fe6ef365-1ec6-4f2c-960c-e2137c920d31"
    },
    "defineProfessorAgent": {
      "context": "chronicle:agent:Prof%20Plum",
      "txId": "ee61a7b1-853c-44fd-a871-27168100bfc5"
    }
  }
}
```

> ### Recording the start of labwork

Reviewer: ```suggestion
### Recording the Start of Labwork
```

> the Labwork activity is started simply with the following mutation

Reviewer: ```suggestion
the `Labwork` activity is started simply with the following mutation
```

```graphql
mutation{
  startActivity(id:{externalId:"Labwork"}){
    context
    txId
  }
}
```

Before any real world work is done an experiment is planned as part of labwork
therefore the experiment entity is generated at the start of the labwork
activity

> ```graphql

Reviewer: maybe prettify the formatting of the code blocks

mutation{
defineExperimentEntity(
externalId: "Chemistry labwork"
attributes: {reactionAttribute: "Suzuki reaction R^1-x + R^2-BY -> R^1-R-2"
reagentsAttribute: "Halide, OrganoBoron species, Pd catalyst, base"
approvalAttribute: "Approved with safety sheet" }
) {
context
txId
}
}
```

OUTPUT

> ```json

Reviewer: Let's prettify the formatting of all code blocks - if you use Altair GraphQL client there's a prettify button you can use.

```graphql
mutation {
  defineResultsEntity(
    externalId: "Chemistry experiment results"
    attributes: {
      rawDataAttribute: "rawdata"
      reportURLAttribute: "reportURL"
      purityAttribute: "98.6"
      yieldAttribute: "5mg"
    }
  ) {
    context
    txId
  }
}
```

OUTPUT

```json
{
  "data": {
    "defineResultsEntity": {
      "context": "chronicle:entity:Chemistry%20experiment%20results",
      "txId": "ff46cac8-1d26-42e1-9722-87d9536cf70f"
    }
  }
}
```

This is then tied to the activity

```graphql
mutation {
  wasGeneratedBy(
    id: { externalId: "Chemistry experiment results" }
    activity: { externalId: "Labwork" }
  ) {
    context
    txId
  }
  wasAssociatedWith(
    activity: { externalId: "Labwork" }
    responsible: { externalId: "Joe Bloggs" }
    role: STUDENT
  ) {
    context
    txId
  }
}
```

OUTPUT

```json
{
  "data": {
    "wasGeneratedBy": {
      "context": "chronicle:entity:Chemistry%20experiment%20results",
      "txId": "842f8f60-d54f-4772-8e37-bb8ac14e609f"
    },
    "wasAssociatedWith": {
      "context": "chronicle:agent:Joe%20Bloggs",
      "txId": "7a8e5f36-791e-49a5-a7d8-65990e2aa074"
    }
  }
}
```

> With the Labwork done and the activity is brought to an end

Reviewer: ```suggestion
With the Labwork done the activity is brought to an end
```

```graphql
mutation{
  endActivity(id:{externalId:"Labwork"}){
    context
    txId
  }
}
```

OUTPUT

```json
{
  "data": {
    "wasGeneratedBy": {
      "context": "chronicle:entity:Chemistry%20experiment%20results",
      "txId": "842f8f60-d54f-4772-8e37-bb8ac14e609f"
    },
    "wasAssociatedWith": {
      "context": "chronicle:agent:Joe%20Bloggs",
      "txId": "7a8e5f36-791e-49a5-a7d8-65990e2aa074"
    }
  }
}
```

> ### Recording the assessment of LabWork

Reviewer: ```suggestion
### Recording the Assessment of LabWork
```

  ) {
    context
    txId
  }
  wasAssociatedWith(
    activity: { externalId: "Assesment" }
    responsible: { externalId: "Prof Plum" }
    role: PROFESSOR
  ) {
    context
    txId
  }
  used(
    id: { externalId: "Chemistry experiment" }
    activity: { externalId: "Labwork" }
  ) {
    context
    txId
  }
}
```

OUTPUT

```json
{
  "data": {
    "wasGeneratedBy": {
      "context": "chronicle:entity:Chemistry%20experiment%20marks",
      "txId": "d8bb0110-4b5f-4f8e-b334-202979aa78b7"
    },
    "wasAssociatedWith": {
      "context": "chronicle:agent:Prof%20Plum",
      "txId": "28f40a32-df56-4605-9ad4-47ec9f355ce2"
    },
    "used": {
      "context": "chronicle:entity:Chemistry%20experiment",
      "txId": "2057d83b-d4e6-416b-9efe-c27db0f73087"
    }
  }
}
```

Then the activity is over and ended with the following mutation

```graphql
mutation {
  endActivity(id: { externalId: "Assesment" }) {
    context
    txId
  }
}
```

OUTPUT

```json
{
  "data": {
    "endActivity": {
      "context": "chronicle:activity:Assesment",
      "txId": "785ed7f1-17dd-42e7-8b9e-f9fb97a9dd6b"
    }
  }
}
```

This concludes a cycle tracking what output a student made and who marked it

> ## Querying the science project

Reviewer: ```suggestion
## Querying the Science Project
```
